### PR TITLE
Adds tests that check partitioning, codegen are reproducible

### PR DESCRIPTION
**Pull request by kaushikcfd**
Draft because:
- [x] Includes commits from #289

### CODE REVIEW

**Comment by inducer** [comment]
Clicked rebase, couldn't help myself. Hope that's OK!

File: test/test_codegen.py [hunk 1]
```diff
@@ -1620,6 +1620,77 @@ def test_zero_size_cl_array_dedup(ctx_factory):
     assert num_nodes_new == (num_nodes_old - 2)
 
 
+# {{{ test_deterministic_codegen
+
+def _get_random_dag_with_comm_nodes(seed):
```
Comment (inducer) [inline test/test_codegen.py:1625]: "comm nodes" sounds like "communication nodes", which I'm taking to mean MPI/distributed. But I don't see any of those here? (Looks like maybe this is a copy-paste from the distributed test? Can those be unified and put into testlib? Does that make sense?)

File: test/test_codegen.py [hunk 2]
```diff
@@ -1620,6 +1620,77 @@ def test_zero_size_cl_array_dedup(ctx_factory):
...
+        seed = 120 + i
+        results = mp_ctx.Array("i", (0, ) * nprocs)
+        print(f"Step {i} {seed}")
+
+        ref_dag = _get_random_dag_with_comm_nodes(seed)
+        ref_t_unit = pt.generate_loopy(ref_dag).program
+
+        # {{{ spawn nprocs-processes and verify they all compare equally
+
+        procs = [mp_ctx.Process(target=_check_deterministic_codegen,
```
Comment (inducer) [inline test/test_codegen.py:1673]: Would it make sense to explicitly set `PYTHONHASHSEED` to different things?

**Review by inducer — COMMENTED** [review]
Thanks! Two questions here ahead of a proper review.

File: test/test_codegen.py (hunk 1 above)
Comment (kaushikcfd) [reply test/test_codegen.py]: Oops, yep. The name was copy-pasted blindly and hence incorrect. (Thanks!)

File: test/test_codegen.py (hunk 2 above)
Comment (kaushikcfd) [reply test/test_codegen.py:1651]: Yep, now we set PYTHONHASHSEED before spawning the processes.

**Review by inducer — APPROVED** [review]
Thanks! LGTM once you've got mypy sorted out.

**Comment by kaushikcfd** [comment]
> Thanks! LGTM once you've got mypy sorted out.

Thanks, fixed! Setting to "Automerge".

**Pull request merged**